### ISSUE TITLE
Too many clippy warnings

### ISSUE THREAD

**Issue by OfficialURL** [dimforge/nalgebra#913]
Currently, running `cargo clippy` on `nalgebra` yields 581 warnings.

A few of these are false positives, and some others are just complaints about the complicated types of `nalgebra` (for which not much can be done). However, many of these should be strictly good improvements to the code. Many of these fixes would also be breaking changes, so we should be aware of that.

I believe that all of these warnings should either be fixed or turned off. That way, if any future sub-par code props up, we'll quickly be able to detect it and fix it.

**Comment by sebcrozet** [comment]
Yeah, `nalgebra` started being developed before `clippy` existed, and I never actually used since then unfortunately. There have been some efforts to fix clippy warning [on other merged PRs](https://github.com/dimforge/nalgebra/pulls?q=is%3Apr+is%3Aclosed+clippy) but obviously we haven't fixed everything yet.

**Comment by Andlon** [comment]
I think this is a really worthwhile endeavor.  Once we have gotten to the "0 baseline" (either by fixing problems or judiciously ignoring individual special cases where we are doing the right thing), we can also consider adding stricter lints on top.

For example, we recently had a case where an automatic derive of `Deserialize` caused an unsoundness hole due to invariants used in unsafe code not being upheld. It turns out that there's a clippy lint (not turned on by default) that would have caught this issue, see https://github.com/RustSec/advisory-db/issues/880 for more information.
